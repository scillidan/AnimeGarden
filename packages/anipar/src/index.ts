import type { ParseOptions, ParseResult } from './types.ts';

export type { ParseResult };

export function parse(title: string, options: ParseOptions = {}): ParseResult | undefined {
  return undefined;
}
