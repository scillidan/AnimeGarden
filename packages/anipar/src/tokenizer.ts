export class Token {
  public readonly text: string;

  public readonly left: string | undefined;

  public readonly right: string | undefined;

  public constructor(text: string, left?: string, right?: string) {
    this.text = text;
    this.left = left;
    this.right = right;
  }

  public get isWrapped() {
    return this.left && this.right;
  }

  public slice(start: number, end?: number) {
    const text = this.text.slice(start, end);
    return new Token(text, this.left, this.right);
  }

  public trim() {
    const text = this.text.trim();
    return new Token(text, this.left, this.right);
  }

  public toString() {
    return `${this.left ?? ''}${this.text}${this.right ?? ''}`;
  }
}

export const Wrappers = new Map([
  ['[', ']'],
  ['【', '】'],
  ['(', ')'],
  ['（', '）'],
  ['{', '}']
]);

export const RevWrappers = new Map([...Wrappers.entries()].map(([k, v]) => [v, k]));

export function tokenize(text: string) {
  const tokens: Token[] = [];

  let cursor = 0;
  let cur = '';
  let left: string | undefined = undefined;
  let right: string | undefined = undefined;
  while (cursor < text.length) {
    const char = text[cursor];
    if (Wrappers.has(char)) {
      if (cur) {
        tokens.push(new Token(cur.trim()));
        cur = '';
      }
      left = char;
      right = Wrappers.get(char)!;
    } else if (left && right && char === right) {
      tokens.push(new Token(cur.trim(), left, right));
      cur = '';
      left = undefined;
      right = undefined;
    } else {
      cur += char;
    }
    cursor += 1;
  }
  if (cur) {
    tokens.push(new Token(cur.trim()));
  }

  return tokens.filter((t) => t.text);
}
