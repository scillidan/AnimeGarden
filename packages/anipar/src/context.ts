import type { ParseOptions, ParseResult } from './types';

import { Token } from './tokenizer';

export class Context {
  // Left consumed tags
  public left = 0;

  // Right consumed tags
  public right = 0;

  public tokens: Token[];

  public options: ParseOptions;

  public result: Partial<ParseResult> = {};

  public tags: string[] = [];

  public constructor(tokens: Token[], options: ParseOptions) {
    this.tokens = tokens;
    this.options = options;
  }

  public update<K1 extends keyof ParseResult>(key: K1, value: ParseResult[K1]) {
    this.result[key] = value;
  }

  public update2<
    K1 extends keyof Required<ParseResult>,
    K2 extends keyof Required<ParseResult>[K1]
  >(key1: K1, key2: K2, value: Required<ParseResult>[K1][K2]) {
    if (!this.result[key1]) {
      // @ts-expect-error
      this.result[key1] = {};
    }
    // @ts-expect-error
    this.result[key1][key2] = value;
  }

  public validate(): ParseResult | undefined {
    return this.result as ParseResult;
  }
}
