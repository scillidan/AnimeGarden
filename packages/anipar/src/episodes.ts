import { Context } from './context';
import { RevWrappers } from './tokenizer';

const WrappedEpisodeRE = /^(?<ep1>\d+)(?:\.(\d))?(?:[vV](\d+))?$|^第(?<ep2>\d+)[集话]$/;
const WrappedMovieRE = /^Movie [vV](\d+)$/;

const EpisodesRange1 = /^(\d+)-(\d+)(?:\s*.*)$/;
const EpisodesRange2 = /^全(\d+)集$/;

export function parseWrappedEpisodes(ctx: Context) {
  const token = ctx.tokens[ctx.right];
  const text = token.text;
  if (token.isWrapped && matchEpiodes(ctx, text)) {
    ctx.right -= 1;
    return true;
  }
  return false;
}

export function matchEpiodes(ctx: Context, text: string) {
  // 1. Single episode
  {
    const res = WrappedEpisodeRE.exec(text);
    if (res) {
      const ep = +(res.groups?.ep1! || res.groups?.ep2!);
      if (!Number.isNaN(ep)) {
        ctx.update2('episode', 'number', ep);
        // 1.5
        if (res[2] && !Number.isNaN(res[2])) {
          ctx.update2('episode', 'numberSub', +res[2]);
        }
        // 3v2
        if (res[3] && !Number.isNaN(res[3])) {
          ctx.update('version', +res[3]);
        }
        return true;
      }
    }
  }
  {
    // [Movie] / [Movie v2]
    const res = WrappedMovieRE.exec(text);
    if (res) {
      if (res[1] && !Number.isNaN(res[1])) {
        ctx.update('version', +res[1]);
      }
      ctx.update('type', 'Movie');
      return true;
    }
  }

  // 2. Episode range
  {
    // 01-26
    const res = EpisodesRange1.exec(text);
    if (res) {
      const from = +res[1];
      const to = +res[2];
      if (!Number.isNaN(from) && !Number.isNaN(to)) {
        ctx.update2('episodeRange', 'from', from);
        ctx.update2('episodeRange', 'to', to);
        return true;
      }
    }
  }
  {
    // 全26集
    const res = EpisodesRange2.exec(text);
    if (res) {
      const to = +res[1];
      if (!Number.isNaN(to)) {
        ctx.update2('episodeRange', 'from', 1);
        ctx.update2('episodeRange', 'to', to);
        return true;
      }
    }
  }

  return false;
}

const SuffixEpisodeRE = /- (\d+)(?:\.(\d))?$/;

export function parseSuffixEpisodes(ctx: Context) {
  if (ctx.result.episode || ctx.result.episodes || ctx.result.episodeRange) {
    return true;
  }

  const token = ctx.tokens[ctx.right];
  const text = token.text;
  {
    const res = SuffixEpisodeRE.exec(text);
    if (res) {
      const ep = +res[1];
      if (!Number.isNaN(ep)) {
        ctx.update2('episode', 'number', ep);

        // 1.5
        if (res[2] && !Number.isNaN(res[2])) {
          ctx.update2('episode', 'numberSub', +res[2]);
        }

        // Remove token suffix
        ctx.tokens[ctx.right] = token.slice(0, text.length - res[0].length).trim();

        return true;
      }
    }
  }
}

// Season
const SuffixRes: Array<[RegExp, (res: RegExpExecArray, ctx: Context) => boolean]> = [
  [
    /(?:S|Season\s?)(\d+)$/,
    (res, ctx) => {
      const season = +res[1];
      if (!Number.isNaN(season)) {
        if (!ctx.result.season?.number || ctx.result.season.number === season) {
          ctx.update2('season', 'number', season);
          return true;
        }
      }
      return false;
    }
  ],
  [
    /(1st|2nd|3rd|[456789]th) Season$/,
    (res, ctx) => {
      const season = Number.parseInt(res[1]);
      if (!Number.isNaN(season)) {
        if (!ctx.result.season?.number || ctx.result.season.number === season) {
          ctx.update2('season', 'number', season);
          return true;
        }
      }
      return false;
    }
  ],
  [
    /第?(\d+)[季期]$/,
    (res, ctx) => {
      const season = +res[1];
      if (!Number.isNaN(season)) {
        if (!ctx.result.season?.number || ctx.result.season.number === season) {
          ctx.update2('season', 'number', season);
          return true;
        }
      }
      return false;
    }
  ],
  [
    /第?((?:[零一二三四五六七八九]十)?[零一二三四五六七八九])[季期]$/,
    (res, ctx) => {
      const text = res[1];
      const map = {
        零: 0,
        一: 1,
        二: 2,
        三: 3,
        四: 4,
        五: 5,
        六: 6,
        七: 7,
        八: 8,
        九: 9
      };
      const base =
        text.length === 2 && text[0] === '十'
          ? 1
          : text.length === 3
            ? map[text[0] as keyof typeof map]!
            : 0;
      const offset = map[text[text.length - 1] as keyof typeof map];
      const season = base * 10 + offset;
      if (!ctx.result.season?.number || ctx.result.season.number === season) {
        ctx.update2('season', 'number', season);
        return true;
      }
      return false;
    }
  ],
  [
    / (\d+)$/,
    (res, ctx) => {
      const season = +res[1];
      if (
        !Number.isNaN(season) &&
        (ctx.result.episode || ctx.result.episodes || ctx.result.episodeRange)
      ) {
        if (!ctx.result.season?.number || ctx.result.season.number === season) {
          ctx.update2('season', 'number', season);
          return true;
        }
      }
      return false;
    }
  ],
  [
    /(?<ep1>\d+)(?:\.(\d))?(?:[vV](\d+))?$|^第(?<ep2>\d+)[集话]$/,
    (res, ctx) => {
      const ep = +(res.groups?.ep1! || res.groups?.ep2!);
      if (!Number.isNaN(ep)) {
        ctx.update2('episode', 'number', ep);
        // 1.5
        if (res[2] && !Number.isNaN(res[2])) {
          ctx.update2('episode', 'numberSub', +res[2]);
        }
        // 3v2
        if (res[3] && !Number.isNaN(res[3])) {
          ctx.update('version', +res[3]);
        }
        return true;
      }
      return false;
    }
  ],
  [
    /（日配版）$/,
    (res, ctx) => {
      ctx.update('language', res[1]);
      return true;
    }
  ]
];

export function parseSuffixSeasonOrEpisodes(ctx: Context, text: string) {
  while (true) {
    let found = false;
    // Ends with: [45]
    if (RevWrappers.has(text[text.length - 1])) {
      const leftWrapper = RevWrappers.get(text[text.length - 1])!;
      const leftIdx = text.lastIndexOf(leftWrapper);
      if (leftIdx !== -1) {
        const maybe = text.slice(leftIdx + 1, text.length - 1);
        if (maybe.length > 0) {
          for (const [re, fn] of SuffixRes) {
            const res = re.exec(maybe);
            if (res && res[0].length === maybe.length && fn(res, ctx)) {
              text = text.slice(0, text.length - maybe.length - 2).trimEnd();
              found = true;
              break;
            }
          }
        }
      }
    }
    for (const [re, fn] of SuffixRes) {
      const res = re.exec(text);
      if (res && fn(res, ctx)) {
        text = text.slice(0, text.length - res[0].length).trimEnd();
        found = true;
        break;
      }
    }
    if (!found) break;
  }
  return text;
}
