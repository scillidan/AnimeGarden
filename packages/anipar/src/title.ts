import { Context } from './context';
import { parseSuffixEpisodes, parseSuffixSeasonOrEpisodes } from './episodes';

export function parseFansub(ctx: Context) {
  // [fansub] title
  if (ctx.left + 1 > ctx.right) {
    return false;
  }

  const token = ctx.tokens[ctx.left];
  if (token.isWrapped) {
    const text = token.text;

    let found = false;
    const seps = ['&', '·'];
    for (const sep of seps) {
      const [name, ...collab] = text.split(sep);

      if (collab.length > 0) {
        // Hack: [jibaketa合成&壓制]
        if (collab.length === 1 && collab[0] === '壓制') {
          break;
        }

        found = true;
        ctx.update2('fansub', 'name', name);
        ctx.update2('fansub', 'collab', collab);
        break;
      }
    }

    if (!found) {
      ctx.update2('fansub', 'name', text);
    }

    ctx.left += 1;

    return true;
  }

  return false;
}

export function parseTitle(ctx: Context) {
  // 1. Parse suffix episodes
  parseSuffixEpisodes(ctx);

  // 2. Parse titles
  const rest = ctx.tokens.slice(ctx.left, ctx.right + 1);
  if (rest.length === 0) return false;

  const text = rest.length === 1 ? rest[0].text : rest.map((t) => t.toString()).join('');

  // 2.1. Try split multiple titles
  let found = false;
  const separators = ['/', '\\'];
  for (const sep of separators) {
    const parts = text
      .split(sep)
      .map((t) => t.trim())
      .filter((t) => !!t);
    if (parts.length > 1) {
      const [title, ...other] = parts;

      const trimmedTitle = parseSuffixSeasonOrEpisodes(ctx, title);
      const trimmedOther = other
        .map((t) => parseSuffixSeasonOrEpisodes(ctx, t))
        .map((t) => t.trim())
        .filter((t) => !!t);

      ctx.update('title', trimmedTitle);
      if (trimmedOther.length > 0) {
        ctx.update('titles', trimmedOther);
      }
      found = true;
    }
  }
  if (!found) {
    const trimmed = parseSuffixSeasonOrEpisodes(ctx, text);
    ctx.update('title', trimmed);
  }

  return true;
}
