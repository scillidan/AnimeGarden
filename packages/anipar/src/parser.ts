import type { ParseOptions, ParseResult } from './types';

import { Context } from './context';
import { tokenize } from './tokenizer';
import { parseLeftTags, parseRightTags } from './keyword';
import { parseFansub, parseTitle } from './title';

export function parse(title: string, options: ParseOptions = {}): ParseResult | undefined {
  const tokens = tokenize(title);
  if (tokens.length === 0) return undefined;

  const context = new Context(tokens, options);

  // 1. Parse right tags
  parseRightTags(context);
  // 2. Parse left tags
  parseLeftTags(context);
  // 3. Parse fansub
  parseFansub(context);
  parseLeftTags(context);
  // 4. Parse title
  if (!parseTitle(context)) {
    return undefined;
  }

  // 6. Return result
  return context.validate();
}
