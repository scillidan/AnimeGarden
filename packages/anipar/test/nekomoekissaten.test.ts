import { describe, it, expect } from 'vitest';

import { parse } from '../src';

const titles = [
  '【喵萌奶茶屋】★剧场版★[欢迎来到驹田蒸馏所 / 駒田蒸留所へようこそ / Komada Jouryuujo e Youkoso][BDRip][1080p][简日双语][招募翻译时轴]',
  '【喵萌奶茶屋】★劇場版★[歡迎來到駒田蒸餾所 / 駒田蒸留所へようこそ / Komada Jouryuujo e Youkoso][BDRip][1080p][繁日雙語][招募翻譯時軸]',
  '【喵萌奶茶屋】★10月新番★[乱马 1/2 2024年版 / Ranma ½ / Ranma 1/2 (2024)][11][1080p][简日双语][招募翻译]',
  '【喵萌奶茶屋】★10月新番★[亂馬 1/2 2024年版 / Ranma ½ / Ranma 1/2 (2024)][11][1080p][繁日雙語][招募翻譯]',
  '【喵萌奶茶屋】★10月新番★[青箱 / Ao no Hako / Blue Box][12][1080p][繁日雙語][招募翻譯]',
  '【喵萌奶茶屋】★10月新番★[Chi。-关于地球的运动- / Chi. Chikyuu no Undou ni Tsuite][11][1080p][简日双语][招募翻译]',
  '【喵萌奶茶屋】★10月新番★[青箱 / Ao no Hako / Blue Box][12][1080p][简日双语][招募翻译]',
  '【喵萌奶茶屋】★10月新番★[妻子變成小學生。 / Tsuma, Shougakusei ni Naru][09][1080p][繁日雙語][招募翻譯時軸]',
  '【喵萌奶茶屋】★10月新番★[妻子变成小学生。 / Tsuma, Shougakusei ni Naru][09][1080p][简日双语][招募翻译时轴]',
  '【喵萌奶茶屋】★10月新番★[超自然武裝噹噠噹 / 膽大黨 / Dandadan][11][1080p][繁日雙語][招募翻譯]',
  '【喵萌奶茶屋】★10月新番★[超自然武装当哒当 / 胆大党 / Dandadan][11][1080p][简日双语][招募翻译]',
  '【喵萌奶茶屋】★10月新番★[神選 / KamiErabi / GOD.app][23][1080p][繁日雙語][招募翻譯]',
  '【喵萌奶茶屋】★10月新番★[神选 / KamiErabi / GOD.app][23][1080p][简日双语][招募翻译]',
  '【喵萌奶茶屋】★07月新番★[亚托莉 我挚爱的时光 / ATRI - My Dear Moments][01-13][1080p][简日双语][招募翻译时轴]',
  '【喵萌奶茶屋】★07月新番★[亞托莉 我摯愛的時光 / ATRI - My Dear Moments][01-13][1080p][繁日雙語][招募翻譯時軸]',
  '【喵萌奶茶屋】★10月新番★[冻牌 / Touhai: Ura Rate Mahjong Touhai Roku][10][1080p][简体][招募翻译]',
  '【喵萌奶茶屋】★07月新番★[神之塔 二期 / Tower of God S2][20][1080p][繁體][招募翻譯]',
  '【喵萌奶茶屋】★10月新番★[乱马 1/2 2024年版 / Ranma ½ / Ranma 1/2 (2024)][10][1080p][简日双语][招募翻译]',
  '【喵萌奶茶屋】★07月新番★[神之塔 二期 / Tower of God S2][22][1080p][繁體][招募翻譯]',
  '【喵萌奶茶屋】★07月新番★[神之塔 二期 / Tower of God S2][21][1080p][繁體][招募翻譯]',
  '【喵萌奶茶屋】★07月新番★[神之塔 二期 / Tower of God S2][20][1080p][简体][招募翻译]',
  '【喵萌奶茶屋】★10月新番★[凍牌 / Touhai: Ura Rate Mahjong Touhai Roku][10][1080p][繁體][招募翻譯]',
  '【喵萌奶茶屋】★07月新番★[神之塔 二期 / Tower of God S2][21][1080p][简体][招募翻译]',
  '【喵萌奶茶屋】★07月新番★[神之塔 二期 / Tower of God S2][22][1080p][简体][招募翻译]',
  '【喵萌奶茶屋】★10月新番★[亂馬 1/2 2024年版 / Ranma ½ / Ranma 1/2 (2024)][10][1080p][繁日雙語][招募翻譯]',
  '【喵萌奶茶屋】★10月新番★[青箱 / Ao no Hako / Blue Box][11][1080p][繁日雙語][招募翻譯]',
  '【喵萌Production】★10月新番★[LoveLive! 超級明星!! 3期 / Love Live! Superstar!! S3][07][1080p][繁日雙語][招募翻譯]'
];

describe('nekomoekissaten', () => {
  for (const title of titles) {
    it(title, () => {
      expect(parse(title)).toMatchSnapshot();
    });
  }
});
