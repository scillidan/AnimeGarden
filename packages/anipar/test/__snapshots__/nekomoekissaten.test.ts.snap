// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`nekomoekissaten > 【喵萌Production】★10月新番★[LoveLive! 超級明星!! 3期 / Love Live! Superstar!! S3][07][1080p][繁日雙語][招募翻譯] 1`] = `
{
  "episode": {
    "number": 7,
  },
  "fansub": {
    "name": "喵萌Production",
  },
  "file": {
    "video": {
      "resolution": "1080p",
    },
  },
  "language": "繁日雙語",
  "month": 10,
  "season": {
    "number": 3,
  },
  "tags": [],
  "title": "LoveLive! 超級明星!!",
  "titles": [
    "Love Live! Superstar!!",
  ],
}
`;

exports[`nekomoekissaten > 【喵萌奶茶屋】★07月新番★[亚托莉 我挚爱的时光 / ATRI - My Dear Moments][01-13][1080p][简日双语][招募翻译时轴] 1`] = `
{
  "episodeRange": {
    "from": 1,
    "to": 13,
  },
  "fansub": {
    "name": "喵萌奶茶屋",
  },
  "file": {
    "video": {
      "resolution": "1080p",
    },
  },
  "language": "简日双语",
  "month": 7,
  "tags": [],
  "title": "亚托莉 我挚爱的时光",
  "titles": [
    "ATRI - My Dear Moments",
  ],
}
`;

exports[`nekomoekissaten > 【喵萌奶茶屋】★07月新番★[亞托莉 我摯愛的時光 / ATRI - My Dear Moments][01-13][1080p][繁日雙語][招募翻譯時軸] 1`] = `
{
  "episodeRange": {
    "from": 1,
    "to": 13,
  },
  "fansub": {
    "name": "喵萌奶茶屋",
  },
  "file": {
    "video": {
      "resolution": "1080p",
    },
  },
  "language": "繁日雙語",
  "month": 7,
  "tags": [],
  "title": "亞托莉 我摯愛的時光",
  "titles": [
    "ATRI - My Dear Moments",
  ],
}
`;

exports[`nekomoekissaten > 【喵萌奶茶屋】★07月新番★[神之塔 二期 / Tower of God S2][20][1080p][简体][招募翻译] 1`] = `
{
  "episode": {
    "number": 20,
  },
  "fansub": {
    "name": "喵萌奶茶屋",
  },
  "file": {
    "video": {
      "resolution": "1080p",
    },
  },
  "language": "简体",
  "month": 7,
  "season": {
    "number": 2,
  },
  "tags": [],
  "title": "神之塔",
  "titles": [
    "Tower of God",
  ],
}
`;

exports[`nekomoekissaten > 【喵萌奶茶屋】★07月新番★[神之塔 二期 / Tower of God S2][20][1080p][繁體][招募翻譯] 1`] = `
{
  "episode": {
    "number": 20,
  },
  "fansub": {
    "name": "喵萌奶茶屋",
  },
  "file": {
    "video": {
      "resolution": "1080p",
    },
  },
  "language": "繁體",
  "month": 7,
  "season": {
    "number": 2,
  },
  "tags": [],
  "title": "神之塔",
  "titles": [
    "Tower of God",
  ],
}
`;

exports[`nekomoekissaten > 【喵萌奶茶屋】★07月新番★[神之塔 二期 / Tower of God S2][21][1080p][简体][招募翻译] 1`] = `
{
  "episode": {
    "number": 21,
  },
  "fansub": {
    "name": "喵萌奶茶屋",
  },
  "file": {
    "video": {
      "resolution": "1080p",
    },
  },
  "language": "简体",
  "month": 7,
  "season": {
    "number": 2,
  },
  "tags": [],
  "title": "神之塔",
  "titles": [
    "Tower of God",
  ],
}
`;

exports[`nekomoekissaten > 【喵萌奶茶屋】★07月新番★[神之塔 二期 / Tower of God S2][21][1080p][繁體][招募翻譯] 1`] = `
{
  "episode": {
    "number": 21,
  },
  "fansub": {
    "name": "喵萌奶茶屋",
  },
  "file": {
    "video": {
      "resolution": "1080p",
    },
  },
  "language": "繁體",
  "month": 7,
  "season": {
    "number": 2,
  },
  "tags": [],
  "title": "神之塔",
  "titles": [
    "Tower of God",
  ],
}
`;

exports[`nekomoekissaten > 【喵萌奶茶屋】★07月新番★[神之塔 二期 / Tower of God S2][22][1080p][简体][招募翻译] 1`] = `
{
  "episode": {
    "number": 22,
  },
  "fansub": {
    "name": "喵萌奶茶屋",
  },
  "file": {
    "video": {
      "resolution": "1080p",
    },
  },
  "language": "简体",
  "month": 7,
  "season": {
    "number": 2,
  },
  "tags": [],
  "title": "神之塔",
  "titles": [
    "Tower of God",
  ],
}
`;

exports[`nekomoekissaten > 【喵萌奶茶屋】★07月新番★[神之塔 二期 / Tower of God S2][22][1080p][繁體][招募翻譯] 1`] = `
{
  "episode": {
    "number": 22,
  },
  "fansub": {
    "name": "喵萌奶茶屋",
  },
  "file": {
    "video": {
      "resolution": "1080p",
    },
  },
  "language": "繁體",
  "month": 7,
  "season": {
    "number": 2,
  },
  "tags": [],
  "title": "神之塔",
  "titles": [
    "Tower of God",
  ],
}
`;

exports[`nekomoekissaten > 【喵萌奶茶屋】★10月新番★[Chi。-关于地球的运动- / Chi. Chikyuu no Undou ni Tsuite][11][1080p][简日双语][招募翻译] 1`] = `
{
  "episode": {
    "number": 11,
  },
  "fansub": {
    "name": "喵萌奶茶屋",
  },
  "file": {
    "video": {
      "resolution": "1080p",
    },
  },
  "language": "简日双语",
  "month": 10,
  "tags": [],
  "title": "Chi。-关于地球的运动-",
  "titles": [
    "Chi. Chikyuu no Undou ni Tsuite",
  ],
}
`;

exports[`nekomoekissaten > 【喵萌奶茶屋】★10月新番★[乱马 1/2 2024年版 / Ranma ½ / Ranma 1/2 (2024)][10][1080p][简日双语][招募翻译] 1`] = `
{
  "episode": {
    "number": 10,
  },
  "fansub": {
    "name": "喵萌奶茶屋",
  },
  "file": {
    "video": {
      "resolution": "1080p",
    },
  },
  "language": "简日双语",
  "month": 10,
  "tags": [],
  "title": "乱马 1/2 2024年版",
  "titles": [
    "Ranma ½",
    "Ranma 1/2(2024)]",
  ],
}
`;

exports[`nekomoekissaten > 【喵萌奶茶屋】★10月新番★[乱马 1/2 2024年版 / Ranma ½ / Ranma 1/2 (2024)][11][1080p][简日双语][招募翻译] 1`] = `
{
  "episode": {
    "number": 11,
  },
  "fansub": {
    "name": "喵萌奶茶屋",
  },
  "file": {
    "video": {
      "resolution": "1080p",
    },
  },
  "language": "简日双语",
  "month": 10,
  "tags": [],
  "title": "乱马 1/2 2024年版",
  "titles": [
    "Ranma ½",
    "Ranma 1/2(2024)]",
  ],
}
`;

exports[`nekomoekissaten > 【喵萌奶茶屋】★10月新番★[亂馬 1/2 2024年版 / Ranma ½ / Ranma 1/2 (2024)][10][1080p][繁日雙語][招募翻譯] 1`] = `
{
  "episode": {
    "number": 10,
  },
  "fansub": {
    "name": "喵萌奶茶屋",
  },
  "file": {
    "video": {
      "resolution": "1080p",
    },
  },
  "language": "繁日雙語",
  "month": 10,
  "tags": [],
  "title": "亂馬 1/2 2024年版",
  "titles": [
    "Ranma ½",
    "Ranma 1/2(2024)]",
  ],
}
`;

exports[`nekomoekissaten > 【喵萌奶茶屋】★10月新番★[亂馬 1/2 2024年版 / Ranma ½ / Ranma 1/2 (2024)][11][1080p][繁日雙語][招募翻譯] 1`] = `
{
  "episode": {
    "number": 11,
  },
  "fansub": {
    "name": "喵萌奶茶屋",
  },
  "file": {
    "video": {
      "resolution": "1080p",
    },
  },
  "language": "繁日雙語",
  "month": 10,
  "tags": [],
  "title": "亂馬 1/2 2024年版",
  "titles": [
    "Ranma ½",
    "Ranma 1/2(2024)]",
  ],
}
`;

exports[`nekomoekissaten > 【喵萌奶茶屋】★10月新番★[冻牌 / Touhai: Ura Rate Mahjong Touhai Roku][10][1080p][简体][招募翻译] 1`] = `
{
  "episode": {
    "number": 10,
  },
  "fansub": {
    "name": "喵萌奶茶屋",
  },
  "file": {
    "video": {
      "resolution": "1080p",
    },
  },
  "language": "简体",
  "month": 10,
  "tags": [],
  "title": "冻牌",
  "titles": [
    "Touhai: Ura Rate Mahjong Touhai Roku",
  ],
}
`;

exports[`nekomoekissaten > 【喵萌奶茶屋】★10月新番★[凍牌 / Touhai: Ura Rate Mahjong Touhai Roku][10][1080p][繁體][招募翻譯] 1`] = `
{
  "episode": {
    "number": 10,
  },
  "fansub": {
    "name": "喵萌奶茶屋",
  },
  "file": {
    "video": {
      "resolution": "1080p",
    },
  },
  "language": "繁體",
  "month": 10,
  "tags": [],
  "title": "凍牌",
  "titles": [
    "Touhai: Ura Rate Mahjong Touhai Roku",
  ],
}
`;

exports[`nekomoekissaten > 【喵萌奶茶屋】★10月新番★[妻子变成小学生。 / Tsuma, Shougakusei ni Naru][09][1080p][简日双语][招募翻译时轴] 1`] = `
{
  "episode": {
    "number": 9,
  },
  "fansub": {
    "name": "喵萌奶茶屋",
  },
  "file": {
    "video": {
      "resolution": "1080p",
    },
  },
  "language": "简日双语",
  "month": 10,
  "tags": [],
  "title": "妻子变成小学生。",
  "titles": [
    "Tsuma, Shougakusei ni Naru",
  ],
}
`;

exports[`nekomoekissaten > 【喵萌奶茶屋】★10月新番★[妻子變成小學生。 / Tsuma, Shougakusei ni Naru][09][1080p][繁日雙語][招募翻譯時軸] 1`] = `
{
  "episode": {
    "number": 9,
  },
  "fansub": {
    "name": "喵萌奶茶屋",
  },
  "file": {
    "video": {
      "resolution": "1080p",
    },
  },
  "language": "繁日雙語",
  "month": 10,
  "tags": [],
  "title": "妻子變成小學生。",
  "titles": [
    "Tsuma, Shougakusei ni Naru",
  ],
}
`;

exports[`nekomoekissaten > 【喵萌奶茶屋】★10月新番★[神选 / KamiErabi / GOD.app][23][1080p][简日双语][招募翻译] 1`] = `
{
  "episode": {
    "number": 23,
  },
  "fansub": {
    "name": "喵萌奶茶屋",
  },
  "file": {
    "video": {
      "resolution": "1080p",
    },
  },
  "language": "简日双语",
  "month": 10,
  "tags": [],
  "title": "神选",
  "titles": [
    "KamiErabi",
    "GOD.app",
  ],
}
`;

exports[`nekomoekissaten > 【喵萌奶茶屋】★10月新番★[神選 / KamiErabi / GOD.app][23][1080p][繁日雙語][招募翻譯] 1`] = `
{
  "episode": {
    "number": 23,
  },
  "fansub": {
    "name": "喵萌奶茶屋",
  },
  "file": {
    "video": {
      "resolution": "1080p",
    },
  },
  "language": "繁日雙語",
  "month": 10,
  "tags": [],
  "title": "神選",
  "titles": [
    "KamiErabi",
    "GOD.app",
  ],
}
`;

exports[`nekomoekissaten > 【喵萌奶茶屋】★10月新番★[超自然武装当哒当 / 胆大党 / Dandadan][11][1080p][简日双语][招募翻译] 1`] = `
{
  "episode": {
    "number": 11,
  },
  "fansub": {
    "name": "喵萌奶茶屋",
  },
  "file": {
    "video": {
      "resolution": "1080p",
    },
  },
  "language": "简日双语",
  "month": 10,
  "tags": [],
  "title": "超自然武装当哒当",
  "titles": [
    "胆大党",
    "Dandadan",
  ],
}
`;

exports[`nekomoekissaten > 【喵萌奶茶屋】★10月新番★[超自然武裝噹噠噹 / 膽大黨 / Dandadan][11][1080p][繁日雙語][招募翻譯] 1`] = `
{
  "episode": {
    "number": 11,
  },
  "fansub": {
    "name": "喵萌奶茶屋",
  },
  "file": {
    "video": {
      "resolution": "1080p",
    },
  },
  "language": "繁日雙語",
  "month": 10,
  "tags": [],
  "title": "超自然武裝噹噠噹",
  "titles": [
    "膽大黨",
    "Dandadan",
  ],
}
`;

exports[`nekomoekissaten > 【喵萌奶茶屋】★10月新番★[青箱 / Ao no Hako / Blue Box][11][1080p][繁日雙語][招募翻譯] 1`] = `
{
  "episode": {
    "number": 11,
  },
  "fansub": {
    "name": "喵萌奶茶屋",
  },
  "file": {
    "video": {
      "resolution": "1080p",
    },
  },
  "language": "繁日雙語",
  "month": 10,
  "tags": [],
  "title": "青箱",
  "titles": [
    "Ao no Hako",
    "Blue Box",
  ],
}
`;

exports[`nekomoekissaten > 【喵萌奶茶屋】★10月新番★[青箱 / Ao no Hako / Blue Box][12][1080p][简日双语][招募翻译] 1`] = `
{
  "episode": {
    "number": 12,
  },
  "fansub": {
    "name": "喵萌奶茶屋",
  },
  "file": {
    "video": {
      "resolution": "1080p",
    },
  },
  "language": "简日双语",
  "month": 10,
  "tags": [],
  "title": "青箱",
  "titles": [
    "Ao no Hako",
    "Blue Box",
  ],
}
`;

exports[`nekomoekissaten > 【喵萌奶茶屋】★10月新番★[青箱 / Ao no Hako / Blue Box][12][1080p][繁日雙語][招募翻譯] 1`] = `
{
  "episode": {
    "number": 12,
  },
  "fansub": {
    "name": "喵萌奶茶屋",
  },
  "file": {
    "video": {
      "resolution": "1080p",
    },
  },
  "language": "繁日雙語",
  "month": 10,
  "tags": [],
  "title": "青箱",
  "titles": [
    "Ao no Hako",
    "Blue Box",
  ],
}
`;

exports[`nekomoekissaten > 【喵萌奶茶屋】★剧场版★[欢迎来到驹田蒸馏所 / 駒田蒸留所へようこそ / Komada Jouryuujo e Youkoso][BDRip][1080p][简日双语][招募翻译时轴] 1`] = `
{
  "fansub": {
    "name": "喵萌奶茶屋",
  },
  "file": {
    "video": {
      "resolution": "1080p",
    },
  },
  "language": "简日双语",
  "source": "BDRip",
  "tags": [],
  "title": "欢迎来到驹田蒸馏所",
  "titles": [
    "駒田蒸留所へようこそ",
    "Komada Jouryuujo e Youkoso",
  ],
  "type": "剧场版",
}
`;

exports[`nekomoekissaten > 【喵萌奶茶屋】★劇場版★[歡迎來到駒田蒸餾所 / 駒田蒸留所へようこそ / Komada Jouryuujo e Youkoso][BDRip][1080p][繁日雙語][招募翻譯時軸] 1`] = `
{
  "fansub": {
    "name": "喵萌奶茶屋",
  },
  "file": {
    "video": {
      "resolution": "1080p",
    },
  },
  "language": "繁日雙語",
  "source": "BDRip",
  "tags": [],
  "title": "歡迎來到駒田蒸餾所",
  "titles": [
    "駒田蒸留所へようこそ",
    "Komada Jouryuujo e Youkoso",
  ],
  "type": "劇場版",
}
`;
